chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,181 @@
+import itertools
+import numpy as np
+
+
+def _count_table(length_all):
+    """Returns a send_cnt matrix.
+
+    The matrix is a 2d numpy array of size NP x NP, where NP is the comm_size,
+    of which element [sender, receiver] is the send/recv count for
+    send()/recv() call.
+
+    i.e.
+    sender rank i sends N elements to receiver j where N = table[i, j]
+    """
+
+    comm_size = len(length_all)
+    table = np.zeros([comm_size, comm_size], dtype=int)
+
+    # Distribute `length_all` elements over all processes
+    # To achieve a good balance, the reminder is scattered over
+    # processes of [rank, rank+1, ..., 0, 1]
+    for send_rank in range(comm_size):
+        sender_length = length_all[send_rank]
+        q = sender_length // comm_size
+        rem = sender_length % comm_size
+        for recv_rank in range(comm_size):
+            if (recv_rank - send_rank + comm_size) % comm_size < rem:
+                table[send_rank, recv_rank] = q + 1
+            else:
+                table[send_rank, recv_rank] = q
+        assert sum(table[send_rank, :]) == sender_length
+    return table
+
+
+def _send_recv_pairs(size):
+    pairs = [(x, y) for x in range(size) for y in range(size)]
+    # Sort the list, so send-recv communications from lower rank to higher rank
+    # happen earlier to avoid deadlock
+    # ex.)  [(0, 0), (1, 1), (0, 1), (1, 0)]
+    #     --> [(0, 0), (0, 1), (1, 0), (1, 1)]
+    pairs.sort()
+    return pairs
+
+
+def _exclusive_scan(array, dtype=None):
+    """C++'s std::exclusive_scan"""
+    if dtype is None:
+        dtype = array.dtype
+    z = np.zeros(len(array), dtype=dtype)
+    z[1:] = np.cumsum(array, dtype=dtype)[:-1]
+
+    return z
+
+
+def _exchange_chunks(comm, chunk, chunk_length_all):
+    count_table = _count_table(chunk_length_all)
```

Comment:
With `scatter_obj` design, calculating this table isn't needed and can be much simpler. Count table would be useful when the send-recv pattern is truly working in parallel.

Reply:
the algorithm is totally refactored.